org.metafacture.fix/src/main/java/org/metafacture/metafix/Metafix.java

```diff
@@ -0,0 +1,292 @@
+        LOG.debug("Start record: {}", identifier);
+        flattener.startRecord(identifier);
+        entityCountStack.clear();
+        entityCount = 0;
+        entityCountStack.add(Integer.valueOf(entityCount));
+        recordIdentifier = identifier;
+        entities = new ArrayList<>();
+        literal(StandardEventNames.ID, identifier);
+    }
+
+    @Override
+    public void endRecord() {
+        entityCountStack.removeLast();
+        if (!entityCountStack.isEmpty()) {
+            throw new IllegalStateException(ENTITIES_NOT_BALANCED);
+        }
+        flattener.endRecord();
+        LOG.debug("End record, walking fix: {}", currentRecord);
+        final RecordTransformer transformer = new RecordTransformer(currentRecord, vars, fix);
+        currentRecord = transformer.transform();
+        if (!currentRecord.containsKey("__reject")) {
+            outputStreamReceiver.startRecord(recordIdentifier);
+            LOG.debug("Sending results to {}", outputStreamReceiver);
+            currentRecord.keySet().stream().filter(k -> !k.startsWith("_")).forEach(k -> {
+                emit(k, currentRecord.get(k));
+            });
+            outputStreamReceiver.endRecord();
+        }
+    }
+
+    private void emit(final Object key, final Object val) {
+        if (val == null) {
+            return;
+        }
+        final List<?> vals = asList(val);
+        final boolean isMulti = vals.size() > 1 || val instanceof List;
+        if (isMulti) {
+            outputStreamReceiver.startEntity(key.toString() + "[]");
+        }
+        for (int i = 0; i < vals.size(); ++i) {
+            final Object value = vals.get(i);
+            if (value instanceof Map) {
+                final Map<?, ?> nested = (Map<?, ?>) value;
+                outputStreamReceiver.startEntity(isMulti ? "" : key.toString());
+                nested.entrySet().forEach(nestedEntry -> {
+                    emit(nestedEntry.getKey(), nestedEntry.getValue());
+                });
+                outputStreamReceiver.endEntity();
+            }
+            else {
+                outputStreamReceiver.literal(isMulti ? (i + 1) + "" : key.toString(), value.toString());
+            }
+        }
+        if (isMulti) {
+            outputStreamReceiver.endEntity();
+        }
+    }
+
+    @Override
+    public void startEntity(final String name) {
+        if (name == null) {
+            throw new IllegalArgumentException("Entity name must not be null.");
+        }
+        ++entityCount;
+        final Integer currentEntityIndex = entityCountStack.peek() - 1;
+        final Map<String, Object> previousEntity = currentEntityIndex < 0 ||
+                entities.size() <= currentEntityIndex ? null : entities.get(currentEntityIndex);
+        entityCountStack.push(Integer.valueOf(entityCount));
+        flattener.startEntity(name);
+        entities.add(currentEntity(name, previousEntity == null && entities.size() >= 0 ? currentRecord : previousEntity));
```

Comment:
What's the purpose of `&& entities.size() >= 0`? (`List.size()` can't be negative.)

Or is it a typo? (`> 0`? `|| entities.size() == 0`?)

Reply:
Right, that makes no sense. I think we don't (no longer?) need that logic there at all.

```suggestion
        entities.add(currentEntity(name, previousEntity != null ? previousEntity : currentRecord));
```

Reply:
Good, I'll (silently) remove it during my refactor for #64 if that's okay with you.

Reply:
:+1: